File: src/test/java/com/nowsprinting/intellij_mob/config/ui/MobSettingsFormTest.java
package com.nowsprinting.intellij_mob.config.ui;

import com.nowsprinting.intellij_mob.config.MobProjectSettings;
import org.junit.Test;

import static org.junit.Assert.*;

public class MobSettingsFormTest {

    @Test
    public void resetEditorFrom_set_wipBranch() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.wipBranch = "mob-session-2";

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertEquals(sut.wipBranch.getText(), "mob-session-2");
    }

    @Test
    public void resetEditorFrom_set_baseBranch() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.baseBranch = "develop";

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertEquals(sut.baseBranch.getText(), "develop");
    }

    @Test
    public void resetEditorFrom_set_remoteName() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.remoteName = "upstream";

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertEquals(sut.remoteName.getText(), "upstream");
    }

    @Test
    public void resetEditorFrom_set_debug() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.debug = true;

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertTrue(sut.debug.isSelected());
    }

    @Test
    public void resetEditorFrom_set_timer() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.timer = 100;

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertEquals(sut.timer.getText(), "100");
    }

    @Test
    public void resetEditorFrom_set_startWithShare() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.startWithShare = true;

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertTrue(sut.startWithShare.isSelected());
    }

    @Test
    public void resetEditorFrom_set_nextAtExpire() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.nextAtExpire = true;

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertTrue(sut.nextAtExpire.isSelected());
    }

    @Test
    public void resetEditorFrom_set_wipCommitMessage() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.wipCommitMessage = "fix";

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertEquals(sut.wipCommitMessage.getText(), "fix");
    }

    @Test
    public void resetEditorFrom_set_nextStay() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.nextStay = true;

        MobSettingsForm sut = new MobSettingsForm();
        sut.resetEditorFrom(settings);
        assertTrue(sut.nextStay.isSelected());
    }

    private MobProjectSettings createDefaultSettings() {
        MobProjectSettings settings = new MobProjectSettings();
        settings.wipBranch = "mob-session";
        settings.baseBranch = "master";
        settings.remoteName = "origin";
        settings.debug = false;
        settings.timer = 10;
        settings.startWithShare = false;
        settings.nextAtExpire = false;
        settings.wipCommitMessage = "mob next [ci-skip]";
        settings.nextStay = false;
        return settings;
    }

    private MobSettingsForm createDefaultForm() {
        MobSettingsForm form = new MobSettingsForm();
        form.wipBranch.setText("mob-session");
        form.baseBranch.setText("master");
        form.remoteName.setText("origin");
        form.debug.setSelected(false);
        form.timer.setText("10");
        form.startWithShare.setSelected(false);
        form.nextAtExpire.setSelected(false);
        form.wipCommitMessage.setText("mob next [ci-skip]");
        form.nextStay.setSelected(false);
        return form;
    }

    @Test
    public void isModified_notModified_false() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        boolean actual = sut.isModified(settings);
        assertFalse(actual);
    }

    @Test
    public void isModified_modified_wipBranch() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.wipBranch.setText("mob-session-2");
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_baseBranch() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.baseBranch.setText("develop");
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_remoteName() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.remoteName.setText("upstream");
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_debug() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.debug.setSelected(true);
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_timer() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.timer.setText("100");
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_startWithShare() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.startWithShare.setSelected(true);
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_nextAtExpire() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.nextAtExpire.setSelected(true);
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_wipCommitMessage() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.wipCommitMessage.setText("fix");
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void isModified_modified_nextStay() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.nextStay.setSelected(true);
        boolean actual = sut.isModified(settings);
        assertTrue(actual);
    }

    @Test
    public void applyEditorTo_modified_wipBranch() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.wipBranch.setText("mob-session-2");
        sut.applyEditorTo(settings);
        assertEquals("mob-session-2", settings.wipBranch);
    }

    @Test
    public void applyEditorTo_modified_baseBranch() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.baseBranch.setText("develop");
        sut.applyEditorTo(settings);
        assertEquals("develop", settings.baseBranch);
    }

    @Test
    public void applyEditorTo_modified_remoteName() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.remoteName.setText("upstream");
        sut.applyEditorTo(settings);
        assertEquals("upstream", settings.remoteName);
    }

    @Test
    public void applyEditorTo_modified_debug() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.debug.setSelected(true);
        sut.applyEditorTo(settings);
        assertTrue(settings.debug);
    }

    @Test
    public void applyEditorTo_modified_timer() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.timer.setText("100");
        sut.applyEditorTo(settings);
        assertEquals(100, settings.timer);
    }

    @Test
    public void applyEditorTo_modified_startWithShare() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.startWithShare.setSelected(true);
        sut.applyEditorTo(settings);
        assertTrue(settings.startWithShare);
    }

    @Test
    public void applyEditorTo_modified_nextAtExpire() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.nextAtExpire.setSelected(true);
        sut.applyEditorTo(settings);
        assertTrue(settings.nextAtExpire);
    }

    @Test
    public void applyEditorTo_modified_wipCommitMessage() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.wipCommitMessage.setText("fix");
        sut.applyEditorTo(settings);
        assertEquals("fix", settings.wipCommitMessage);
    }

    @Test
    public void applyEditorTo_modified_nextStay() {
        MobProjectSettings settings = createDefaultSettings();
        MobSettingsForm sut = createDefaultForm();
        sut.nextStay.setSelected(true);
        sut.applyEditorTo(settings);
        assertTrue(settings.nextStay);
    }
}

File: src/test/kotlin/com/nowsprinting/intellij_mob/config/ui/MobSettingsFormTest.kt
package com.nowsprinting.intellij_mob.config.ui

import com.nowsprinting.intellij_mob.config.MobProjectSettings
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class MobSettingsFormTest {

    @Test
    fun resetEditorFrom_set_wipBranch() {
        val settings = MobProjectSettings()
        settings.wipBranch = "mob-session-2"
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertEquals(sut.wipBranch.text, "mob-session-2")
    }

    @Test
    fun resetEditorFrom_set_baseBranch() {
        val settings = MobProjectSettings()
        settings.baseBranch = "develop"
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertEquals(sut.baseBranch.text, "develop")
    }

    @Test
    fun resetEditorFrom_set_remoteName() {
        val settings = MobProjectSettings()
        settings.remoteName = "upstream"
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertEquals(sut.remoteName.text, "upstream")
    }

    @Test
    fun resetEditorFrom_set_debug() {
        val settings = MobProjectSettings()
        settings.debug = true
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertTrue(sut.debug.isSelected)
    }

    @Test
    fun resetEditorFrom_set_timer() {
        val settings = MobProjectSettings()
        settings.timer = 100
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertEquals(sut.timer.text, "100")
    }

    @Test
    fun resetEditorFrom_set_startWithShare() {
        val settings = MobProjectSettings()
        settings.startWithShare = true
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertTrue(sut.startWithShare.isSelected)
    }

    @Test
    fun resetEditorFrom_set_nextAtExpire() {
        val settings = MobProjectSettings()
        settings.nextAtExpire = true
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertTrue(sut.nextAtExpire.isSelected)
    }

    @Test
    fun resetEditorFrom_set_wipCommitMessage() {
        val settings = MobProjectSettings()
        settings.wipCommitMessage = "fix"
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertEquals(sut.wipCommitMessage.text, "fix")
    }

    @Test
    fun resetEditorFrom_set_nextStay() {
        val settings = MobProjectSettings()
        settings.nextStay = true
        val sut = MobSettingsForm()
        sut.resetEditorFrom(settings)
        Assertions.assertTrue(sut.nextStay.isSelected)
    }

    private fun createDefaultSettings(): MobProjectSettings {
        val settings = MobProjectSettings()
        settings.wipBranch = "mob-session"
        settings.baseBranch = "master"
        settings.remoteName = "origin"
        settings.debug = false
        settings.timer = 10
        settings.startWithShare = false
        settings.nextAtExpire = false
        settings.wipCommitMessage = "mob next [ci-skip]"
        settings.nextStay = false
        return settings
    }

    private fun createDefaultForm(): MobSettingsForm {
        val form = MobSettingsForm()
        form.wipBranch.text = "mob-session"
        form.baseBranch.text = "master"
        form.remoteName.text = "origin"
        form.debug.isSelected = false
        form.timer.text = "10"
        form.startWithShare.isSelected = false
        form.nextAtExpire.isSelected = false
        form.wipCommitMessage.text = "mob next [ci-skip]"
        form.nextStay.isSelected = false
        return form
    }

    @Test
    fun isModified_notModified_false() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        val actual = sut.isModified(settings)
        Assertions.assertFalse(actual)
    }

    @Test
    fun isModified_modified_wipBranch() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.wipBranch.text = "mob-session-2"
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_baseBranch() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.baseBranch.text = "develop"
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_remoteName() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.remoteName.text = "upstream"
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_debug() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.debug.isSelected = true
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_timer() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.timer.text = "100"
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_startWithShare() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.startWithShare.isSelected = true
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_nextAtExpire() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.nextAtExpire.isSelected = true
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_wipCommitMessage() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.wipCommitMessage.text = "fix"
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun isModified_modified_nextStay() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.nextStay.isSelected = true
        val actual = sut.isModified(settings)
        Assertions.assertTrue(actual)
    }

    @Test
    fun applyEditorTo_modified_wipBranch() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.wipBranch.text = "mob-session-2"
        sut.applyEditorTo(settings)
        Assertions.assertEquals("mob-session-2", settings.wipBranch)
    }

    @Test
    fun applyEditorTo_modified_baseBranch() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.baseBranch.text = "develop"
        sut.applyEditorTo(settings)
        Assertions.assertEquals("develop", settings.baseBranch)
    }

    @Test
    fun applyEditorTo_modified_remoteName() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.remoteName.text = "upstream"
        sut.applyEditorTo(settings)
        Assertions.assertEquals("upstream", settings.remoteName)
    }

    @Test
    fun applyEditorTo_modified_debug() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.debug.isSelected = true
        sut.applyEditorTo(settings)
        Assertions.assertTrue(settings.debug)
    }

    @Test
    fun applyEditorTo_modified_timer() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.timer.text = "100"
        sut.applyEditorTo(settings)
        Assertions.assertEquals(100, settings.timer)
    }

    @Test
    fun applyEditorTo_modified_startWithShare() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.startWithShare.isSelected = true
        sut.applyEditorTo(settings)
        Assertions.assertTrue(settings.startWithShare)
    }

    @Test
    fun applyEditorTo_modified_nextAtExpire() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.nextAtExpire.isSelected = true
        sut.applyEditorTo(settings)
        Assertions.assertTrue(settings.nextAtExpire)
    }

    @Test
    fun applyEditorTo_modified_wipCommitMessage() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.wipCommitMessage.text = "fix"
        sut.applyEditorTo(settings)
        Assertions.assertEquals("fix", settings.wipCommitMessage)
    }

    @Test
    fun applyEditorTo_modified_nextStay() {
        val settings = createDefaultSettings()
        val sut = createDefaultForm()
        sut.nextStay.isSelected = true
        sut.applyEditorTo(settings)
        Assertions.assertTrue(settings.nextStay)
    }
}